File: session-05-chaos-engineering/go/resilience/retry.go
package resilience

import (
	"fmt"
	"time"

	"coffeeshop-chaos/kitchen"
)

// Exercise 3: Retry wrapper. Retries up to MaxAttempts times.
type RetryKitchenService struct {
	Inner       kitchen.KitchenService
	MaxAttempts int
	Delay       time.Duration
}

func (s *RetryKitchenService) Prepare(orderID int, drink string) (*kitchen.Preparation, error) {
	var lastErr error
	for attempt := 1; attempt <= s.MaxAttempts; attempt++ {
		prep, err := s.Inner.Prepare(orderID, drink)
		if err == nil {
			if attempt > 1 {
				fmt.Printf("[Retry] Succeeded on attempt %d for order #%d\n", attempt, orderID)
			}
			return prep, nil
		}
		lastErr = err
		fmt.Printf("[Retry] Attempt %d/%d failed for order #%d: %s\n", attempt, s.MaxAttempts, orderID, err)
		if attempt < s.MaxAttempts {
			time.Sleep(s.Delay)
		}
	}
	return nil, fmt.Errorf("all %d attempts failed for order #%d: %w", s.MaxAttempts, orderID, lastErr)
}

File: session-05-chaos-engineering/python/resilience/retry.py
import time

from kitchen.service import KitchenService, Preparation


class RetryKitchenService(KitchenService):
    """Exercise 3: Retries up to max_attempts times on failure."""

    def __init__(self, inner: KitchenService, max_attempts: int = 3, delay_seconds: float = 1.0):
        self.inner = inner
        self.max_attempts = max_attempts
        self.delay_seconds = delay_seconds

    def prepare(self, order_id: int, drink: str) -> Preparation:
        last_error = None
        for attempt in range(1, self.max_attempts + 1):
            try:
                result = self.inner.prepare(order_id, drink)
                if attempt > 1:
                    print(f"[Retry] Succeeded on attempt {attempt} for order #{order_id}")
                return result
            except Exception as e:
                last_error = e
                print(f"[Retry] Attempt {attempt}/{self.max_attempts} failed for order #{order_id}: {e}")
                if attempt < self.max_attempts:
                    time.sleep(self.delay_seconds)
        raise RuntimeError(f"All {self.max_attempts} attempts failed for order #{order_id}") from last_error
